keps/sig-architecture/20191001-prevent-permabeta.md

```diff
@@ -0,0 +1,163 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+k8s.io REST APIs should not languish in beta.  They should take feedback and progress towards GA by either
+1. meeting GA criteria and getting promoted, or
+2. having a new beta and deprecating the previous beta
+  
+This must happen within nine months (three releases).  If it does not,
+the REST API will be deprecated with an announced intent to remove the API per the [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+## Motivation
+
+When a REST API reaches beta, it is turned on by default.  This is great for getting feedback, but it can also lead to state
+where users and vendors start building important infrastructure against APIs that are not considered stable.
+In addition, once a REST API is on by default, the incentive to further stabilize appears to diminish.
+See the REST API that have been beta for a long time: CSRs and Ingresses as examples.
+If we're honest with ourselves, a single actor has been cleaning up behind a lot of the project to unstick perma-beta APIs.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+1. Prevent k8s.io REST APIs from being in a single beta version for more than nine months.
+2. Prevent beta APIs from being treated as GA by users and vendors.
+
+### Non-Goals
+
+1. Promote APIs to GA before they are ready.
+2. Control non-k8s.io REST APIs.
+3. Control features that are not REST APIs.
+4. Control fields on otherwise GA REST APIs.
+
+## Proposal
+
+Once a REST API reaches beta, it has nine months to 
+1. reach GA and deprecate the beta or 
+2. have a new beta version and deprecate the previous beta.
```

Comment:
To clarify - in other words you're saying that after nine months we will deprecate this beta API no matter what. What differs is what happens in addition to this, which may be:
- graduation to GA
- having another beta
- nothing

Assuming that I'm not missing anything, maybe we should explicitly say that each beta will be deprecated after 9 months + another 9 months of deprecation policy means will be removed after 1.5 years.
Or am I missing something?

[For me being more explicit about plans is better.]

Reply:
Do we need a new beta version number or is this a "re-authorize" beta?